src/business_logic/transaction/objects/internal_declare.rs

```diff
@@ -0,0 +1,428 @@
+            self.sender_address.clone(),
+            self.hash_value.clone(),
+            self.signature.clone(),
+            self.max_fee,
+            self.nonce.clone(),
+            n_steps,
+            self.version,
+        )
+    }
+
+    pub fn run_validate_entrypoint<S: Default + State + StateReader>(
+        &self,
+        state: &mut S,
+        resources_manager: &mut ExecutionResourcesManager,
+        general_config: &StarknetGeneralConfig,
+    ) -> Result<Option<CallInfo>, ExecutionError> {
+        if self.version > 0x8000_0000_0000_0000 {
+            return Ok(None);
+        }
+
+        let calldata = self.get_calldata();
+
+        let entry_point = ExecutionEntryPoint::new(
+            self.account_contract_address(),
+            calldata,
+            self.validate_entry_point_selector.clone(),
+            Address(Felt::zero()),
+            EntryPointType::External,
+            None,
+            None,
+        );
+
+        let call_info = entry_point.execute(
+            state,
+            general_config,
+            resources_manager,
+            &self.get_execution_context(general_config.invoke_tx_max_n_steps),
+        )?;
+
+        verify_no_calls_to_other_contracts(&call_info)
+            .map_err(|_| ExecutionError::UnauthorizedActionOnValidate)?;
+
+        Ok(Some(call_info))
+    }
+
+    // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    // Calculates and charges the actual fee.
+    // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+    pub fn charge_fee<S: Default + State + StateReader + Clone>(
+        &self,
+        state: &mut S,
+        resources: HashMap<String, usize>,
+        general_config: StarknetGeneralConfig,
+    ) -> Result<FeeInfo, TransactionError> {
+        if self.max_fee.is_zero() {
+            return Ok((None, 0));
+        }
+
+        let actual_fee = calculate_tx_fee(
+            resources,
+            general_config.starknet_os_config.gas_price,
+            &general_config,
+        )?;
+
+        let tx_context = self.get_execution_context(general_config.invoke_tx_max_n_steps);
+        let fee_transfer_info =
+            execute_fee_transfer(state, general_config, tx_context, actual_fee)?;
+
+        Ok((Some(fee_transfer_info), actual_fee))
+    }
+
+    fn handle_nonce<S: Default + State + StateReader + Clone>(
+        &self,
+        state: &mut S,
+    ) -> Result<(), TransactionError> {
+        if self.version > 0x8000_0000_0000_0000 {
+            return Err(TransactionError::StarknetError(
+                "Don't handle nonce for version 0".to_string(),
+            ));
+        }
+
+        let contract_address = self.account_contract_address();
+        let current_nonce = state.get_nonce_at(&contract_address)?.to_owned();
+        if current_nonce != self.nonce {
+            return Err(TransactionError::InvalidTransactionNonce(
+                current_nonce.to_string(),
+                self.nonce.clone().to_string(),
```

Comment:
I'm pretty sure that `.to_string()` doesn't take ownership of the value (therefore you can remove the `.clone()`).

Reply:
indeed.